packages/fetch-engine/src/getLatestTag.ts

```diff
@@ -0,0 +1,209 @@
+      'introspection-engine',
+      'migration-engine',
+      'prisma-fmt',
+    ]) {
+      for (const extension of [
+        '.gz',
+        '.gz.sha256',
+        '.gz.sig',
+        '.sig',
+        '.sha256',
+      ]) {
+        const downloadUrl = getDownloadUrl(
+          branch,
+          commit,
+          platform,
+          engine,
+          extension,
+        )
+        urls.push(downloadUrl)
+      }
+    }
+  }
+
+  return urls
+}
+
+async function getFirstFinishedCommit(
+  branch: string,
+  commits: string[],
+): Promise<string | void> {
+  for (const commit of commits) {
+    const urls = getAllUrls(branch, commit)
+    // TODO: potential to speed things up
+    // We don't always need to wait for the last commit
+    const exist = await pMap(urls, urlExists, { concurrency: 10 })
+    const hasMissing = exist.some((e) => !e)
+
+    if (!hasMissing) {
+      return commit
+    } else {
+      const missing = urls.filter((_, i) => !exist[i])
+      // if all are missing, we don't have to talk about it
+      // it might just be a broken commit or just still building
+      if (missing.length !== urls.length) {
+        console.log(
+          `${chalk.blueBright(
+            'info',
+          )} The engine commit ${commit} is not yet done. We're skipping it as we're in dev. Missing urls: ${
+            missing.length
+          }`,
+        )
+      }
+    }
+  }
+}
+
+export async function urlExists(url) {
+  try {
+    const res = await fetch(url, {
+      method: 'HEAD',
+      agent: getProxyAgent(url) as any,
+    })
+
+    const headers = fromEntries(res.headers.entries())
+    if (res.status > 200) {
+      // console.error(res, headers)
+    }
+    if (parseInt(headers['content-length']) > 0) {
+      return res.status < 300
+    }
+  } catch (e) {
+    //
+    // console.error(e)
+  }
+  return false
+}
+
+function fromEntries<T>(
+  entries: IterableIterator<[string, T]>,
+): Record<string, T> {
+  const result = {}
+  for (const [key, value] of entries) {
+    result[key] = value
+  }
+  return result
+}
+
+async function getBranch() {
+  if (process.env.NODE_ENV !== 'test') {
+    if (process.env.PATCH_BRANCH) {
+      return process.env.PATCH_BRANCH
+    }
+    if (process.env.BUILDKITE_BRANCH) {
+      return process.env.BUILDKITE_BRANCH
+    }
+    if (process.env.GITHUB_CONTEXT) {
+      const context = JSON.parse(process.env.GITHUB_CONTEXT)
+      return context.head_ref
+    }
+  }
+
+  // Need to be handled
+  // for example it's used in https://github.com/prisma/binary-tester and the environment
+  // is not a git repository so it fails
+  try {
+    const result = await execa.command('git rev-parse --abbrev-ref HEAD', {
+      shell: true,
+      stdio: 'pipe',
+    })
+    return result.stdout
+  } catch (e) {
+    console.error(e)
+  }
+
+  return
+}
+
+// TODO: Adjust this for stable release
+function isPatchBranch(version: string): boolean {
+  return /^2\.(\d+)\.x/.test(version)
```

Comment:
Oh hey I know that code - we should also make this major safe if we can already.